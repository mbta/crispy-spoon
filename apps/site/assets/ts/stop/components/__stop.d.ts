import {
  Direction,
  DirectionId,
  EnhancedRoute,
  Stop,
  Alert
} from "../../__v3api";
import { MapData } from "../../leaflet/components/__mapdata";

export interface TypedRoutes {
  group_name: string;
  routes: RouteWithDirections[];
}

export interface StopMapData {
  map_data: MapData;
  map_srcset: string;
  map_url: string;
}

export interface RouteWithDirection {
  direction_id: DirectionId | null;
  route: EnhancedRoute;
}

export interface SuggestedTransfer {
  stop: Stop;
  distance: number;
  routes_with_direction: RouteWithDirection[];
}

export interface StopPageData {
  stop: Stop;
  street_view_url: string | null;
  routes: TypedRoutes[];
  routes_with_direction?: RouteWithDirection[];
  tabs: Tab[];
  zone_number: string;
  alerts: Alert[];
  suggested_transfers: SuggestedTransfer[];
  retail_locations: RetailLocationWithDistance[];
  tab: string;
  alerts_tab: AlertsTab;
  routes_and_alerts: { [key: string]: Alert[] };
}

interface AlertData {
  alerts: Alert[];
  empty_message: string;
}

export interface AlertsTab {
  [key: string]: AlertData | string;
  upcoming: AlertData;
  current: AlertData;
  all: AlertData;
  initial_selected: string;
}

export interface Tab {
  badge?: TabBadge;
  class: string;
  href: string;
  id: string;
  name: string;
  selected: string;
}

export interface TabBadge {
  content: string;
  aria_label: string;
  class?: string;
}

export interface RetailLocation {
  address: string;
  latitude: number;
  longitude: number;
  name: string;
  payment: string[];
  phone: string;
}

export interface RetailLocationWithDistance {
  distance: string;
  location: RetailLocation;
}

export interface RouteWithDirections {
  route: EnhancedRoute;
  directions: Direction[];
}
