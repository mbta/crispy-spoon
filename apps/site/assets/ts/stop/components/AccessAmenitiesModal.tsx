import React from "react";
import { Alert, Facility } from "../../__v3api";
import Alerts from "../../components/Alerts";
import { hasFacilityAlert } from "../../models/alert";
import { AmenityModal, AmenityLink } from "./amenities/AmenityCard";

const AccessAmenitiesModal = ({
  stopName,
  alerts,
  facilities,
  facilityType
}: {
  stopName: string;
  alerts: Alert[];
  facilities: Facility[];
  facilityType: "Elevators" | "Escalators";
}): JSX.Element => {
  const hasFacilities = facilities ? facilities.length > 0 : false;
  const linkText =
    facilityType === "Elevators"
      ? "Report an elevator issue"
      : "Report an escalator issue";
  return (
    <>
      {hasFacilities && (
        <AmenityModal headerText={`${facilityType} at ${stopName}`}>
          <Alerts alerts={alerts} />
          <h2>Elevator Status</h2>

          <table className="access-amenities-table">
            <tbody>
              <tr className="access-amenities-header access-amenities-row">
                <th className="access-amenities-name">Elevator</th>
                <th className="status">Status</th>
              </tr>
              {facilities?.map(facility => {
                return (
                  <tr
                    key={facility.id}
                    className="access-amenities-row access-amenities-text"
                  >
                    <td className="access-amenities-name">
                      {facility.attributes.short_name}
                    </td>
                    {hasFacilityAlert(facility.id, alerts) ? (
                      <td className="status">
                        <i className="fa-solid fa-circle amenity-status amenity-out" />
                        Out of Order
                      </td>
                    ) : (
                      <td className="status">
                        <i className="fa-solid fa-circle amenity-status amenity-working" />
                        Working
                      </td>
                    )}
                  </tr>
                );
              })}
            </tbody>
          </table>
          <AmenityLink
            url="/accessibility/trip-planning"
            text="Plan an accessible trip"
          />
          <AmenityLink url="/customer-support" text={linkText} />
        </AmenityModal>
      )}
    </>
  );
};
export default AccessAmenitiesModal;
