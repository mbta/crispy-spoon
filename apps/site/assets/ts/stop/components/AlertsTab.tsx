import React, { useState, ReactElement } from "react";
import { AlertsTab as AlertsTabType, AlertData } from "./__stop";
import Alerts from "../../components/Alerts";
import { handleReactEnterKeyPress } from "../../helpers/keyboard-events";

interface Props {
  alertsTab: AlertsTabType;
}

const selectedClass = (filter: string, selected: string): string =>
  selected === filter ? "m-alerts__time-filter--selected" : "";

const timeFilter = (
  name: string,
  title: string,
  currentFilter: string,
  onClick: React.MouseEventHandler
): ReactElement<HTMLElement> => (
  <div
    role="button"
    id={name}
    tabIndex={0}
    className={`m-alerts__time-filter ${selectedClass(name, currentFilter)}`}
    onClick={onClick}
    onKeyPress={e => handleReactEnterKeyPress(e, onClick)}
  >
    {title}
  </div>
);

const AlertsTab = ({ alertsTab }: Props): ReactElement<HTMLElement> => {
  // eslint-disable-next-line camelcase
  const initialFilter: string = alertsTab.initial_selected || "all";
  const [filter, changeFilter] = useState(initialFilter);

  return (
    <div className="container">
      <div className="page-section station">
        <div className="page-section">
          <div className="row">
            <div className="col-xs-12 col-lg-3">
              <h3 className="h3">Filter by type</h3>
              <div className="m-alerts__time-filters">
                {timeFilter("all", "All Alerts", filter, () =>
                  changeFilter("all")
                )}
                {timeFilter("current", "Current Alerts", filter, () =>
                  changeFilter("current")
                )}
                {timeFilter("upcoming", "Planned Service Alerts", filter, () =>
                  changeFilter("upcoming")
                )}
              </div>
            </div>
            <div className="col-xs-12 col-lg-8 col-lg-offset-1">
              <h2 className="h2">Alerts</h2>
              <div>
                {/* eslint-disable */}
                {(alertsTab[filter] as AlertData).alerts.length ? (
                  <Alerts alerts={(alertsTab[filter] as AlertData).alerts} />
                ) : (
                  <div className="callout">
                    {(alertsTab[filter] as AlertData).empty_message}
                  </div>
                )}
                {/* eslint-enable */}
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default AlertsTab;
