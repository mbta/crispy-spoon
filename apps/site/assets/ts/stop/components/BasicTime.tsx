import React from "react";
import parse from "date-fns/parse";
import { formatRelativeTime, formatToBostonTime } from "../../helpers/date";

export interface BasicTimeProps {
  displayType: "relative" | "absolute";
  strikethrough?: boolean;
  time: Date | undefined;
  targetDate?: Date;
}
const BasicTime = ({
  displayType,
  strikethrough,
  time,
  targetDate
}: BasicTimeProps): JSX.Element | null => {
  if (!time) return null;

  const formattedTime =
    displayType === "relative"
      ? formatRelativeTime(time, targetDate)
      : formatToBostonTime(time, "h:mm aa");

  // displayType === "relative" might produce an absolute-formatted time if it's
  // more than an hour away. parse the result to find the truth.
  const parsedTime = parse(formattedTime, "h:mm aa", new Date()).getTime();
  const isAbsolute = Number.isNaN(parsedTime) ? "relative" : "absolute";
  // add data attribute to be used for later manipulation and filtering
  const dataType: "relative" | "absolute" | "imminent" =
    formattedTime === "<1 minute away" ? "imminent" : isAbsolute;
  return (
    <time
      data-type={dataType}
      dateTime={time.toISOString()}
      className={strikethrough ? "strikethrough" : ""}
    >
      {formattedTime}
    </time>
  );
};

export default BasicTime;
