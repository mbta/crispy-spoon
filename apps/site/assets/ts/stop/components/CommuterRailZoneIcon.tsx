import React, { ReactElement } from "react";
import { clickRoutePillAction, Dispatch } from "../state";

const crZone = (
  zoneNumber?: string,
  dispatch?: Dispatch
): ReactElement<HTMLElement> | false => {
  const content = `Zone ${zoneNumber}`;

  let tag = (
    <div className="m-stop-page__header-feature">
      <span className="m-stop-page__icon c-icon__cr-zone">{content}</span>
    </div>
  );

  if (dispatch) {
    tag = (
      <a
        href="#route-card-list"
        onClick={() =>
          dispatch && dispatch(clickRoutePillAction("commuter_rail"))
        }
        className="m-stop-page__header-feature"
      >
        <span className="m-stop-page__icon c-icon__cr-zone">{content}</span>
      </a>
    );
  }

  return !!zoneNumber && zoneNumber.length > 0 && tag;
};

const CommuterRailZoneIcon = ({
  zoneNumber,
  dispatch
}: {
  zoneNumber?: string;
  dispatch?: Dispatch;
}): ReactElement<HTMLElement> => {
  return <>{crZone(zoneNumber, dispatch)}</>;
};

export { CommuterRailZoneIcon as default };
