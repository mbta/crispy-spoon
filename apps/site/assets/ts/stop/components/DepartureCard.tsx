import { reject } from "lodash";
import React, { ReactElement } from "react";
import { Alert, Route } from "../../__v3api";
import { routeName, routeToModeIcon } from "../../helpers/route-headers";
import { routeBgClass } from "../../helpers/css";
import renderSvg from "../../helpers/render-svg";
import useDepartureRow from "../../hooks/useDepartureRow";
import { DepartureInfo } from "../../models/departureInfo";
import { allAlertsForDirection } from "../../models/alert";
import { departureInfoInRoutePatterns } from "../../helpers/departureInfo";
import { isACommuterRailRoute, isSubwayRoute } from "../../models/route";
import DepartureTimes from "./DepartureTimes";
import {
  RoutePatternGroup,
  RoutePatternWithPolyline,
  sortedGroupedRoutePatterns
} from "../../models/route-patterns";

const isNoncanonicalAndNoDepartures = (
  routePatterns: RoutePatternWithPolyline[],
  departures: DepartureInfo[]
): boolean => {
  const isNonCanonical = !routePatterns.find(rp => !!rp.canonical);
  return isNonCanonical && departures.length === 0;
};

const DepartureCard = ({
  alertsForRoute,
  departuresForRoute,
  routePatternsByHeadsign,
  route
}: {
  alertsForRoute: Alert[];
  departuresForRoute: DepartureInfo[];
  routePatternsByHeadsign: RoutePatternGroup;
  route: Route;
}): ReactElement<HTMLElement> | null => {
  const { setRow } = useDepartureRow([route]);

  let sortedRoutePatternsByHeadsign = sortedGroupedRoutePatterns(
    routePatternsByHeadsign
  );

  if (isSubwayRoute(route)) {
    // remove certain headsigns based on route pattern and departures
    sortedRoutePatternsByHeadsign = reject(
      sortedRoutePatternsByHeadsign,
      entry => {
        const [, { route_patterns: routePatterns }] = entry;
        const departures = departuresForRoute.filter(d =>
          departureInfoInRoutePatterns(d, routePatterns)
        );
        return isNoncanonicalAndNoDepartures(routePatterns, departures);
      }
    );
    // don't render a route card if there's no headsigns left to show
    if (sortedRoutePatternsByHeadsign.length === 0) return null;
  }

  return (
    <li className="departure-card">
      <a
        className={`departure-card__route ${routeBgClass(route)} notranslate`}
        href={`/schedules/${route.id}`}
        data-turbolinks="false"
      >
        {renderSvg("c-svg__icon", routeToModeIcon(route), true)}{" "}
        {routeName(route)}
      </a>
      {sortedRoutePatternsByHeadsign.map(
        ([
          headsign,
          { direction_id: directionId, route_patterns: routePatterns }
        ]) => {
          const onClick = (): void =>
            setRow({
              routeId: route.id,
              directionId: directionId.toString(),
              headsign
            });

          return (
            <DepartureTimes
              key={headsign}
              alertsForDirection={allAlertsForDirection(
                alertsForRoute,
                directionId
              )}
              headsign={headsign}
              departures={departuresForRoute.filter(d =>
                departureInfoInRoutePatterns(d, routePatterns)
              )}
              onClick={onClick}
              isCR={isACommuterRailRoute(route)}
              hasService={routePatterns.length !== 0}
            />
          );
        }
      )}
    </li>
  );
};

export default DepartureCard;
