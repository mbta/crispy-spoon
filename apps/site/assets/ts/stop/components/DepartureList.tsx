import React, { ReactElement } from "react";
import { filter } from "lodash";
import { Alert, DirectionId, Route, Stop, Trip } from "../../__v3api";
import { DepartureInfo } from "../../models/departureInfo";
import { SUBWAY, departuresListFromInfos } from "../../helpers/departureInfo";
import { routeBgClass } from "../../helpers/css";
import { routeName, routeToModeIcon } from "../../helpers/route-headers";
import renderSvg from "../../helpers/render-svg";
import { isSuppressiveAlert } from "../../models/alert";
import Alerts from "../../components/Alerts";
import { isACommuterRailRoute } from "../../models/route";

interface DepartureListProps {
  route: Route;
  stop: Stop;
  departures: DepartureInfo[];
  directionId: DirectionId;
  headsign: string;
  alerts: Alert[];
  hasService: boolean;
  targetDate?: Date | undefined;
}

const displayNoUpcomingTrips = (
  message = "No more trips today"
): JSX.Element => {
  return (
    <div className="c-alert-item--low m-8 d-flex justify-content-center align-items-center pb-40 pt-40">
      {message}
    </div>
  );
};

const DepartureList = ({
  route,
  stop,
  departures,
  directionId,
  headsign,
  alerts,
  hasService,
  targetDate
}: DepartureListProps): ReactElement<HTMLElement> => {
  const isCR = isACommuterRailRoute(route);

  // don's show cancelled departures for subway
  const modeSpecificDepartures: DepartureInfo[] = filter(
    departures,
    (d: DepartureInfo) => !(d.isCancelled && d.routeMode === SUBWAY)
  );
  const numPredictions = modeSpecificDepartures.filter(d => !!d.prediction)
    .length;
  const alertsShouldSuppressDepartures = alerts.some(alert =>
    isSuppressiveAlert(alert, numPredictions)
  );
  const tripForSelectedRoutePattern: Trip | undefined =
    modeSpecificDepartures[0]?.trip;

  const noTrips =
    modeSpecificDepartures.length === 0 && displayNoUpcomingTrips();
  const noService = !hasService && displayNoUpcomingTrips("No trips today");
  const noServiceOrNoTrips = noService || noTrips;

  return (
    <>
      <div className="stop-departures departure-list-header">
        <div className={`departure-card__route ${routeBgClass(route)}`}>
          <div className="notranslate">
            {renderSvg("c-svg__icon", routeToModeIcon(route), true)}{" "}
            {routeName(route)}
          </div>
          <a
            className="open-schedule"
            href={`/schedules/${route.id}/line?schedule_direction[direction_id]=${directionId}&schedule_direction[variant]=${tripForSelectedRoutePattern?.route_pattern_id}&schedule_finder[direction_id]=${directionId}&schedule_finder[origin]=${stop.id}`}
            data-turbolinks="false"
          >
            View all schedules
          </a>
        </div>
      </div>
      <h2 className="departure-list__sub-header">
        <div className="departure-list__origin-stop-name notranslate">
          {stop.name} to
        </div>
        <div className="departure-list__headsign notranslate">{headsign}</div>
      </h2>
      {alerts.length ? <Alerts alerts={alerts} /> : null}
      {noServiceOrNoTrips ||
        (!alertsShouldSuppressDepartures && (
          <ul className="stop-routes__departures list-unstyled">
            {departuresListFromInfos(modeSpecificDepartures, isCR, targetDate)}
          </ul>
        ))}
    </>
  );
};

export default DepartureList;
