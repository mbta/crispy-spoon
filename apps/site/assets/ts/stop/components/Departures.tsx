import React, { ReactElement, useState } from "react";
import { Stop, Mode, Alert } from "../../__v3api";
import { Dispatch, clickModeAction } from "../state";
import { TypedRoutes, RouteWithDirections } from "./__stop";
import RouteCard from "./RouteCard";
import { ModeFilter, modeByV3ModeType } from "../../components/ModeFilter";

interface Props {
  routes: TypedRoutes[];
  routesAndAlerts?: { [key: string]: Alert[] };
  stop: Stop;
  selectedModes: Mode[];
  dispatch: Dispatch;
}

const includesMultipleModes = (typedRoutes: TypedRoutes[]): boolean =>
  typedRoutes.length > 1;

const allRoutes = (typedRoutes: TypedRoutes[]): RouteWithDirections[] =>
  typedRoutes.reduce(
    (acc: RouteWithDirections[], typeAndRoutes: TypedRoutes) =>
      acc.concat(typeAndRoutes.routes),
    []
  );

const availableModes = (typedRoutes: TypedRoutes[]): string[] =>
  // eslint-disable-next-line camelcase
  typedRoutes.map(({ group_name }) => group_name);

const filteredByModes = (
  routes: RouteWithDirections[],
  modes: Mode[]
): RouteWithDirections[] => {
  // If there are no selections or all selections, do not filter
  if (modes.length === 0 || modes.length === 3) return routes;

  return routes.filter(route =>
    modes.some(mode => modeByV3ModeType[route.route.type] === mode)
  );
};

// Only return the first 2 routes for small screens, or all of them if there are filters, showMore
const filterForSmallScreen = (
  routes: RouteWithDirections[],
  modes: Mode[],
  allRoutesShown: boolean
): RouteWithDirections[] => {
  if (modes.length > 0 || allRoutesShown) {
    return routes;
  }
  return routes.slice(0, 2);
};

const Departures = ({
  routes,
  routesAndAlerts,
  stop,
  selectedModes,
  dispatch
}: Props): ReactElement<HTMLElement> => {
  const [allRoutesShown, setAllRoutesShown] = useState<boolean>(false);
  const isModeSelected = (mode: Mode): boolean => selectedModes.includes(mode);

  const handleModeClick = (mode: Mode): void => dispatch(clickModeAction(mode));

  const filteredRoutes = filteredByModes(allRoutes(routes), selectedModes);
  // Only show the first 2 elements of the array (unless filtered or button clicked)
  const shortFilteredRoutes = filterForSmallScreen(
    filteredRoutes,
    selectedModes,
    allRoutesShown
  );

  const displayShowMoreButton = (
    allRoutesBoolean: boolean,
    modes: Mode[],
    routesArray: RouteWithDirections[]
  ): boolean => {
    return !allRoutesBoolean && modes.length === 0 && routesArray.length > 2;
  };

  return (
    <div id="route-card-list">
      <h2>Departures</h2>

      <div className="m-stop-page__departures">
        {includesMultipleModes(routes) && (
          <ModeFilter
            isModeSelected={isModeSelected}
            onModeClickAction={handleModeClick}
            modeButtonsToShow={availableModes(routes)}
          />
        )}
        <div className="hidden-sm-down mt-8">
          {filteredRoutes.map(routeWithDirections => (
            <RouteCard
              key={routeWithDirections.route.id}
              route={routeWithDirections.route}
              directions={routeWithDirections.directions}
              stop={stop}
              alerts={
                routesAndAlerts
                  ? routesAndAlerts[routeWithDirections.route.id]
                  : []
              }
            />
          ))}
        </div>
        <div className="hidden-md-up mt-8">
          {/* This should only show on small screens */}
          {shortFilteredRoutes.map(routeWithDirections => (
            <RouteCard
              key={routeWithDirections.route.id}
              route={routeWithDirections.route}
              directions={routeWithDirections.directions}
              stop={stop}
              alerts={
                routesAndAlerts
                  ? routesAndAlerts[routeWithDirections.route.id]
                  : []
              }
            />
          ))}
          {displayShowMoreButton(
            allRoutesShown,
            selectedModes,
            filteredRoutes
          ) && (
            <button
              className="btn btn-block mt-10 btn-secondary"
              onClick={() => setAllRoutesShown(true)}
              type="button"
            >
              Show {filteredRoutes.length - 2} more routes
            </button>
          )}
        </div>
      </div>
    </div>
  );
};

export default Departures;
