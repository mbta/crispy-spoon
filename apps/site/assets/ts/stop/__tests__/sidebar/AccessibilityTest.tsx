import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import stopData from "../stopData.json";
import Accessibility from "../../components/sidebar/Accessibility";
import { Stop } from "../../../__v3api";
import { StopPageData, TypedRoutes } from "../../components/__stop";
import { createReactRoot } from "../../../app/helpers/testUtils";

const data = JSON.parse(JSON.stringify(stopData)) as StopPageData;
/* eslint-disable camelcase */

const stop: Stop = {
  type: "station",
  "station?": true,
  parking_lots: [],
  note: null,
  name: "South Station",
  bike_storage: [],
  fare_facilities: [],
  longitude: -71.055242,
  latitude: 42.352271,
  "is_child?": false,
  id: "place-sstat",
  "has_fare_machine?": true,
  "has_charlie_card_vendor?": false,
  closed_stop_info: null,
  address: "700 Atlantic Ave, Boston, MA 02110",
  municipality: "Boston",
  accessibility: [
    "accessible",
    "escalator_both",
    "elevator",
    "fully_elevated_platform"
  ]
};

describe("Accessibility", () => {
  it("it renders", () => {
    createReactRoot();
    const tree = renderer
      .create(
        <Accessibility
          routes={data.routes}
          stop={stop}
          isExpanded
          isFocused={false}
          dispatch={() => {}}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("it renders without extra data on accessible", () => {
    createReactRoot();
    const stopWithoutFeatures = { ...stop, accessibility: ["accessible"] };
    const wrapper = mount(
      <Accessibility
        routes={data.routes}
        stop={stopWithoutFeatures}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    expect(wrapper.find("#accessibility-level").text()).toContain(
      "South Station is accessible."
    );
    expect(wrapper.find("#accessibility-features").exists()).toBeFalsy();
  });

  it("it renders with unknown accessibility", () => {
    createReactRoot();
    const stopWithoutFeatures = { ...stop, accessibility: ["unknown"] };
    const wrapper = mount(
      <Accessibility
        routes={data.routes}
        stop={stopWithoutFeatures}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    expect(wrapper.find("#accessibility-level").text()).toContain(
      "Minor to moderate accessibility barriers exist at South Station. Bus operator may need to relocate bus for safe boarding and exiting."
    );
    expect(wrapper.find("#accessibility-features").exists()).toBeFalsy();
  });

  it("it renders with no information about accessibility data", () => {
    createReactRoot();
    const stopWithoutFeatures = { ...stop, accessibility: [] };
    const wrapper = mount(
      <Accessibility
        routes={data.routes}
        stop={stopWithoutFeatures}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    expect(wrapper.find("#accessibility-level").text()).toContain(
      "Significant accessibility barriers exist at South Station. Customers using wheeled mobility devices may need to board at street level. Bus operator will need to relocate bus for safe boarding and exiting."
    );
    expect(wrapper.find("#accessibility-features").exists()).toBeFalsy();
  });

  it("it renders with information about a partially inaccessible stop", () => {
    createReactRoot();
    const stopWithoutFeatures = { ...stop, accessibility: ["tty_phone"] };
    const wrapper = mount(
      <Accessibility
        routes={data.routes}
        stop={stopWithoutFeatures}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    expect(wrapper.find("#accessibility-level").text()).toContain(
      "Significant accessibility barriers exist at South Station. Customers using wheeled mobility devices may need to board at street level. Bus operator will need to relocate bus for safe boarding and exiting."
    );
    expect(wrapper.find("#accessibility-features").exists()).toBeTruthy();
  });

  it("it renders special case for Ashmont/Mattapan Trolley", () => {
    createReactRoot();
    const stopAshmont = { ...stop, id: "place-asmnl" };
    const wrapper = mount(
      <Accessibility
        routes={data.routes}
        stop={stopAshmont}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    expect(wrapper.find("#accessibility-level").text()).toContain(
      "Significant accessibility barriers exist at South Station but customers can board or exit the Mattapan Trolley using a mobile lift."
    );
    expect(wrapper.find("#accessibility-features").text()).toContain(
      "Escalator"
    );
  });

  it("it renders other special cases", () => {
    createReactRoot();
    const stopAshmont = { ...stop, id: "place-lech" };
    const wrapper = mount(
      <Accessibility
        routes={data.routes}
        stop={stopAshmont}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    expect(wrapper.find("#accessibility-level").text()).toContain(
      "Significant accessibility barriers exist at South Station but customers can board or exit the train using a mobile lift."
    );
    expect(wrapper.find("#accessibility-features").text()).toContain(
      "Escalator"
    );
  });

  it("it handles unknown accessibility types", () => {
    createReactRoot();
    const stopNewFeature = {
      ...stop,
      accessibility: ["accessible", "unknown_accessible_feature"]
    };
    const wrapper = mount(
      <Accessibility
        routes={data.routes}
        stop={stopNewFeature}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    expect(wrapper.find("#accessibility-features").text()).toContain(
      "unknown accessible feature"
    );
  });

  it("it handles types where the accessibility level is not accessible (lacks 'accessible' or 'unknown')", () => {
    createReactRoot();
    const stopNoLevel = {
      ...stop,
      accessibility: ["escalator_up"]
    };
    const wrapper = mount(
      <Accessibility
        routes={data.routes}
        stop={stopNoLevel}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    expect(wrapper.find("#accessibility-level").text()).toContain(
      "Significant accessibility barriers exist"
    );
    expect(wrapper.find("#accessibility-features").exists()).toBeTruthy();
  });

  it("it handles types where the accessibility level is not accessible (lacks 'accessible' or 'unknown') without any bus routes", () => {
    createReactRoot();
    const routesNoBus: TypedRoutes[] = [
      {
        group_name: "subway",
        routes: [
          {
            route: {
              type: 1,
              name: "Orange Line",
              header: "Orange Line",
              long_name: "Orange Line",
              id: "Orange",
              direction_names: {
                "0": "South",
                "1": "North"
              },
              direction_destinations: {
                "0": "Ashmont/Braintree",
                "1": "Alewife"
              },
              description: "rapid_transit",
              alerts: []
            },
            directions: []
          }
        ]
      }
    ];
    const stopNoLevel = {
      ...stop,
      accessibility: ["escalator_up"]
    };
    const wrapper = mount(
      <Accessibility
        routes={routesNoBus}
        stop={stopNoLevel}
        isExpanded
        isFocused={false}
        dispatch={() => {}}
      />
    );
    const accessLevel = wrapper.find("#accessibility-level").text();
    expect(accessLevel).toContain("Significant accessibility barriers exist");
    expect(accessLevel).not.toContain("Bus");
    expect(wrapper.find("#accessibility-features").exists()).toBeTruthy();
  });
});
