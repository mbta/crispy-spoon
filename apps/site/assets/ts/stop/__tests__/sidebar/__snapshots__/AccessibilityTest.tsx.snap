// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accessibility it renders 1`] = `
Array [
  <h3
    aria-controls="panel-accessibility"
    aria-expanded={true}
    className="c-expandable-block__header "
    id="header-accessibility"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span
      aria-hidden="true"
      className="notranslate c-expandable-block__header-icon"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
    <div
      className="c-expandable-block__header-text"
    >
      Accessibility
      <span
        className="fa c-expandable-block__header-caret"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#xF106;",
          }
        }
      />
    </div>
  </h3>,
  <div
    aria-labelledby="header-accessibility"
    className="c-expandable-block__panel "
    id="panel-accessibility"
    role="region"
    tabIndex={0}
  >
    <p
      id="accessibility-level"
    >
      South Station is accessible. 
    </p>
    <div
      id="accessibility-features"
    >
      It 
      has the following features:
      <ul
        className="m-stop-page__sidebar-list"
      >
        <li
          className="c-unordered-list-item"
        >
          Escalator (up and down)
        </li>
        <li
          className="c-unordered-list-item"
        >
          Elevator
        </li>
        <li
          className="c-unordered-list-item"
        >
          Full high level platform to provide level boarding to every car in a train set
        </li>
      </ul>
    </div>
    <p>
      <a
        className="c-call-to-action"
        href="/customer-support?comments=MBTA | Accessibility issue "
      >
        Report an elevator, escalator, or other accessibility issue
      </a>
    </p>
    <p>
      <a
        className="c-call-to-action"
        href="/accessibility"
      >
        Learn more about accessibility
      </a>
    </p>
  </div>,
  <noscript>
    <style>
      #header-accessibility { display: none; }
    </style>
    <h3
      className="c-expandable-block__header"
    >
      <span
        aria-hidden="true"
        className="notranslate c-expandable-block__header-icon"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Accessibility
      <span
        className="fa c-expandable-block__header-caret"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#xF106;",
          }
        }
      />
    </h3>
    <div
      className="c-expandable-block__panel"
      role="region"
    >
      <p
        id="accessibility-level"
      >
        South Station is accessible. 
      </p>
      <div
        id="accessibility-features"
      >
        It 
        has the following features:
        <ul
          className="m-stop-page__sidebar-list"
        >
          <li
            className="c-unordered-list-item"
          >
            Escalator (up and down)
          </li>
          <li
            className="c-unordered-list-item"
          >
            Elevator
          </li>
          <li
            className="c-unordered-list-item"
          >
            Full high level platform to provide level boarding to every car in a train set
          </li>
        </ul>
      </div>
      <p>
        <a
          className="c-call-to-action"
          href="/customer-support?comments=MBTA | Accessibility issue "
        >
          Report an elevator, escalator, or other accessibility issue
        </a>
      </p>
      <p>
        <a
          className="c-call-to-action"
          href="/accessibility"
        >
          Learn more about accessibility
        </a>
      </p>
    </div>
  </noscript>,
]
`;
